docs: initial "helia-fetch" design (#1)
ipfs/web3-fetch: README.md

### Usage

```ts
import { CreateWeb3Fetch } from '@helia/web3-fetch'

> const web3Fetch = await CreateWeb3Fetch({

Author: @lidel yea, I think the consensus on the call last week was that we should call it verified-fetch =) I like that one

  gateways: ['mygateway.info', 'trustless-gateway.link']
})

> const resp = await web3Fetch('ipfs://bafy...')

Reviewer: Do we support `'ipns://QmFoo...'` as well?

const json = await resp.json()
```

### Configuration

#### Usage with customized Helia

You can see variations of Helia and js-libp2p configuration options at https://helia.io/interfaces/helia.index.HeliaInit.html.

The `@helia/http` module is currently in-progress, but the init options should be a subset of the `helia` module's init options. See https://github.com/ipfs/helia/issues/289 for more information.

```ts
import { CreateWeb3Fetch } from '@helia/web3-fetch'
import { CreateHelia as CreateHeliaHttpOnly } from '@helia/http'

> const web3Fetch = await CreateWeb3Fetch({
>   helia: CreateHeliaHttpOnly({
>     routers: ['delegated-ipfs.dev'],
>     gateways: ['mygateway.info', 'trustless-gateway.link'],
>   })
> })

Reviewer: I'd either take a preconfigured Helia node, or `{ gateways: string[], routers: string[] }` as an argument.

This creates an obvious error that should stop people doing things like `{ gateways: string[], routers: string[], helia: Helia }`.

```suggestion
const web3Fetch = await CreateWeb3Fetch(
  CreateHeliaHttpOnly({
    routers: ['delegated-ipfs.dev'],
    gateways: ['mygateway.info', 'trustless-gateway.link'],
  })
)
```

const resp = await web3Fetch('ipfs://bafy...')

> const json = await resp.json()

Reviewer: Something I've never liked about `fetch` is that there's no way of getting any progress information.

A common complaint against IPFS is "I don't know why it's failing", which is why we added the `onProgress` option which lets users have some context-specific feedback as to how far their request got.

I'm not sure how to accommodate this and remain compatible with the `fetch` API.  Do we care that it's a black box?

Reviewer 2: We don't have to be 100% `fetch` compatible, this is not drop-in replacement, it is more about leveraging conventions and level of abstraction that lives in web developer minds.

Perhaps IPFS-specific `onProgress` reporting callback could be passed as one of options?
(`ipfsFetch(resource, options)`)

```

### Comparison to fetch

First, this library will require instantiation in order to configure the gateways, delegated routers, and potentially a custom Helia instance. Secondly, once your web3-fetch method is created, it will act as similar to the `fetch()` API as possible.

[The `fetch()` API](https://developer.mozilla.org/en-US/docs/Web/API/fetch) takes two parameters:

1. A [resource](https://developer.mozilla.org/en-US/docs/Web/API/fetch#resource)
2. An [options object](https://developer.mozilla.org/en-US/docs/Web/API/fetch#options)

#### Resource argument

This library intends to support the following methods of fetching web3 content from IPFS:

> 1. IPFS protocol: `ipfs://<cidv1>` & `ipfs://<cidv2>`

Reviewer 3: Do we want to support UnixFS pathing at all, e.g. `ipfs://cidv1/path/to/my/cat/image`?

What implications would this have?

Reviewer 4: I'd expect this matches the gateway specification (i.e. cid-v2 doesn't exist, but any valid `/ip(f|n)s/*` should be valid as `ip(f|n)s://*`)

Reviewer 4: I'd expect support for CIDv0 and CIDv1 to both be supported.

Author: > I'd expect support for CIDv0 and CIDv1 to both be supported.

+1... omg

Reviewer 2: Hm.. to be aligned with other web browser efforts, we most likely want helia/web-fetch to implement a minimal set of things listed in "Web Pathing Spec" (draft: https://github.com/ipfs/specs/pull/453, feedback welcome).

So yes, dag-cbor+unixfs path traversal, cid0+1, a few common hash functions listed in the spec.

Author: I believe i've addressed concerns here with [`5675a62` (#1)](https://github.com/ipfs/web3-fetch/pull/1/commits/5675a626604d5479759b2cf83d70830d41585ff8)

#### Options argument

This library will not plan to support the exact Fetch API options object, as some of the arguments don't make sense. Instead, it will only support options necessary to meet [IPFS specs](https://specs.ipfs.tech/) related to specifying the resultant shape of desired content.

Some of those header specifications are:

1. https://specs.ipfs.tech/http-gateways/path-gateway/#request-headers
2. https://specs.ipfs.tech/http-gateways/trustless-gateway/#request-headers
3. https://specs.ipfs.tech/http-gateways/subdomain-gateway/#request-headers

Where possible, options and Helia internals will be automatically configured to the appropriate codec & content type based on the `web3-fetch` configuration and `options` argument passed.

Known options that will be supported:

1. `signal` - An AbortSignal that a user can use to abort the request.
2. `redirect` - A string that specifies the redirect type. One of `follow`, `error`, or `manual`. Defaults to `follow`. Best effort to adhere to the [Fetch API redirect](https://developer.mozilla.org/en-US/docs/Web/API/fetch#redirect) parameter.

> 3. `headers` - An object of headers to be sent with the request. Best effort to adhere to the [Fetch API headers](https://developer.mozilla.org/en-US/docs/Web/API/fetch#headers) parameter.

Reviewer 2: Probably main (only) thing to support here will be `Accept` header to allow for requesting responses in trustless [block](https://www.iana.org/assignments/media-types/application/vnd.ipld.raw), [car](https://www.iana.org/assignments/media-types/application/vnd.ipld.car), and [ipns record](https://www.iana.org/assignments/media-types/application/vnd.ipfs.ipns-record) content types.

5. `body` - An object that specifies the body of the request. Best effort to adhere to the [Fetch API body](https://developer.mozilla.org/en-US/docs/Web/API/fetch#body) parameter.
6. `cache` - Will basically act as `force-cache` for the request. Best effort to adhere to the [Fetch API cache](https://developer.mozilla.org/en-US/docs/Web/API/fetch#cache) parameter.

Some in-flight specs (IPIPs) that will affect the options object this library supports in the future can be seen at https://specs.ipfs.tech/ipips, but a few that I'm aware of are:

1. [IPIP-0412: Signaling Block Order in CARs on HTTP Gateways](https://specs.ipfs.tech/ipips/ipip-0412/)
2. [IPIP-0402: Partial CAR Support on Trustless Gateways](https://specs.ipfs.tech/ipips/ipip-0402/)
3. [IPIP-0386: Subdomain Gateway Interop with _redirects](https://specs.ipfs.tech/ipips/ipip-0386/)
4. [IPIP-0328: JSON and CBOR Response Formats on HTTP Gateways](https://specs.ipfs.tech/ipips/ipip-0328/)
5. [IPIP-0288: TAR Response Format on HTTP Gateways](https://specs.ipfs.tech/ipips/ipip-0288/)

#### Response types

This library's purpose is to return reasonable fetch-like responses, which means the following content types:

> 1. Images, videos, audio, etc. as `Blob`

Reviewer: Do we want to add `DAG-JSON` and `DAG-CBOR` here too?

Author: I think I need to be less specific here, and somehow state that fetching content is intended for leaf-node content -- such as images/videos/audio & other assets, or other IPLD content (with link) -- that can be represented by https://developer.mozilla.org/en-US/docs/Web/API/Response#instance_methods

> 2. JSON as `JSON`

Reviewer: Is this returning JSON parsed into a JavaScript object (like `fetch` does) or actually returning a JSON (which is a string)?

Author: At least for the first iteration, if the response is not `JSON.parse`-able we should allow it to throw, as a typical HttpResponse would. If the underlying content is dag-json/dag-cbor/dag-* then it would require  more manipulation and understanding of the content, and should instead follow recommendations discussed at https://github.com/ipfs/web3-fetch/pull/1#discussion_r1448179531

> 3. CBOR as `ArrayBuffer`

Reviewer: Would this be better parsing the CBOR and returning a JavaScript object similar to `.json()` on the `fetch` API?

4. TAR as `Blob`

> 5. CAR as `Blob`

Reviewer: The [Blob constructor](https://developer.mozilla.org/en-US/docs/Web/API/Blob/Blob) requires all parts to be present - this will prevent streaming CAR files to the CAR reader - would this be better consumed as a `ReadableStream`?

Author: 100%. I think all responses will end up being added to the HTTP Response as a stream since data we obtain larger than a single block is pretty much all an asyncIterable. Plus, a stream should be more effective memory-wise. Users would then handle the response as appropriate.

I think marking responses as blobs here is incorrect, but I wanted to get something out ASAP for feedback. 

It would be more accurate to specify that regular [HTTP Response methods](https://developer.mozilla.org/en-US/docs/Web/API/Response#instance_methods) should be used as appropriate: 
An HTTP response can be:
* converted via ".json()" or throw
* converted via ".blob()" almost always, though potentially useless (https://developer.mozilla.org/en-US/docs/Web/API/Response/blob)
* converted via ".arrayBuffer()" 
* converted via ".text()"

Reviewer 2: Library should not make decisions per content type – that is up to the developer that uses it, and we want to follow conventions/abstractions from Fetch API where possible:
- If they want JSON and abstract the differences between dag-json/dag-cbor/json/and json-file-on-unixfs, they would call `.json()` and get JSON object.
- If they want to have entire image to put it in the DOM, use `.blob()` or `.arrayBuffer()`
- If they want to read plain text in utf-8, they would call `.text()`
- If they prefer to stream response data, use something like `response.body.getReader()`

6. All streamable content as a response with a [`ReadableStream`](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API/Using_readable_streams#consuming_a_fetch_as_a_stream)

> If your content doesn't have a mime-type or an [IPFS spec](https://specs.ipfs.tech), this library will not support it.

Reviewer 2: May be more clear if we efer to https://github.com/ipfs/specs/pull/453 here and remove mime-type mention (unixfs supports arbitrary blobs, i think maybe you meant unsupported codec/hashes here?).

Author: I think the point I was trying to make here is that if there's no recognize-able mime-type, or IPFS Spec for some custom type (CAR files), then the response data will likely unusable without more IPFS tools available via Helia.

A question that comes up is if we cannot determine a mime-type, what do we set the Content-Type header to? In [helia-service-worker-gateway](https://github.com/ipfs-shipyard/helia-service-worker-gateway/blob/1ff2c01703f9ecd78fdefebf172378c5e2a71b2f/src/lib/heliaFetch/getContentType.ts#L34-L52) we default to text/plain.. but we need to define and communicate a better default for this lib

Author: Even if unixfs does support arbitrary blobs, does fetching an arbitrary blob make sense for a `@helia/verified-fetch` library? I imagine folks using this library will want useable content, not pieces of blocks or arbitrary block data. If they want those pieces, won't they need further pieces, and therefore `helia` would be a better fit for their usecase?

I imagine this is something we can move forward with for now and address later, sticking to more reasonable web-based use-cases for now (i.e. usable content we can expect the majority of web apps to use today).

We should decide on at least a content-type/mime-type/media-type to set if none can be determined. 

> Warning: Browsers use the MIME type, not the file extension, to determine how to process a URL, so it's important that web servers send the correct MIME type in the response's [Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) header. If this is not correctly configured, browsers are likely to misinterpret the contents of files, sites will not work correctly, and downloaded files may be mishandled.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types

#### Response headers

This library will set the [HTTP Response](https://developer.mozilla.org/en-US/docs/Web/API/Response) headers to the appropriate values for the content type according to the appropriate [IPFS Specifications](https://specs.ipfs.tech/).

Some known header specifications:

* https://specs.ipfs.tech/http-gateways/path-gateway/#response-headers
* https://specs.ipfs.tech/http-gateways/trustless-gateway/#response-headers
* https://specs.ipfs.tech/http-gateways/subdomain-gateway/#response-headers

##### Unsupported response types

> * Returning IPLD nodes or DAGs are not supported, as there is no currently well-defined structure for representing this data in an [HTTP Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). Instead, users should use the [`helia`](https://github.com/ipfs/helia) library directly for this use case.

Reviewer 2: > there is no currently well-defined structure for representing this data in an [HTTP Response]

We have `aplication/vnd.ipld.car` for this, perhaps rephrase:

```suggestion
* Returning IPLD nodes or DAGs as JS objects is not supported, as there is no currently well-defined structure for representing this data in an [HTTP Response](https://developer.mozilla.org/en-US/docs/Web/API/Response). Instead, users should request `aplication/vnd.ipld.car` or use the [`helia`](https://github.com/ipfs/helia) library directly for this use case.
```

### Usage

```ts

> import { CreateVerifiedFetch } from '@helia/verified-fetch'

Author: ```suggestion
import { createVerifiedFetch } from '@helia/verified-fetch'
```

> const web3Fetch = await CreateVerifiedFetch({

Author: ```suggestion
const verifiedFetch = await createVerifiedFetch({
```

  gateways: ['mygateway.info', 'trustless-gateway.link']
})

const resp = await web3Fetch('ipfs://bafy...')

const json = await resp.json()
```

### Configuration

#### Usage with customized Helia

You can see variations of Helia and js-libp2p configuration options at https://helia.io/interfaces/helia.index.HeliaInit.html.

> The `@helia/http` module is currently in-progress, but the init options should be a subset of the `helia` module's init options. See https://github.com/ipfs/helia/issues/289 for more information.

Author: @achingbrain @lidel 

Another thought here is that we may need to modify `@helia/http` or `@helia/block-brokers` if we want to do more than just `vnd.ipld.raw` content fetching. We would need some way to pass headers through to `@helia/http`, or we could simply handle this inside of `@helia/verified-fetch`. 

I think we should handle this inside of `@helia/verified-fetch`.

```ts

> import { CreateVerifiedFetch } from '@helia/verified-fetch'

Author: ```suggestion
import { createVerifiedFetch } from '@helia/verified-fetch'
```

import { CreateHelia as CreateHeliaHttpOnly } from '@helia/http'

> const web3Fetch = await CreateVerifiedFetch(

Author: ```suggestion
const verifiedFetch = await createVerifiedFetch(
```

  CreateHeliaHttpOnly({
    routers: ['delegated-ipfs.dev'],
    gateways: ['mygateway.info', 'trustless-gateway.link'],
  })
)

const resp = await web3Fetch('ipfs://bafy...')

const json = await resp.json()
```

### Comparison to fetch

First, this library will require instantiation in order to configure the gateways and delegated routers, or potentially a custom Helia instance. Secondly, once your verified-fetch method is created, it will act as similar to the `fetch()` API as possible.

[The `fetch()` API](https://developer.mozilla.org/en-US/docs/Web/API/fetch) takes two parameters:

1. A [resource](https://developer.mozilla.org/en-US/docs/Web/API/fetch#resource)
2. An [options object](https://developer.mozilla.org/en-US/docs/Web/API/fetch#options)

#### Resource argument

This library intends to support the following methods of fetching web3 content from IPFS:

1. IPFS protocol: `ipfs://<cidv0>` & `ipfs://<cidv0>`
2. IPNS protocol: `ipns://<peerId>` & `ipns://<publicKey>` & `ipns://<hostUri_Supporting_DnsLink_TxtRecords>`
3. CID instances: An actual CID instance `CID.parse('bafy...')`

> 4. CID strings: A CID string `bafy...`

Author: @lidel @aschmahmann @achingbrain My question here is whether we should start to normalize the usage of the `ipfs://` & `ipns://` protocols, instead of taking a string or CID instance.

Reviewer 2: I'd say the library should accept all, but use `ipfs://` in examples
